File: src/classifier.py
"""PicNote image classifier — determines if an image is informational or casual."""

from __future__ import annotations

import logging
import re
import subprocess

from .watcher import PhotoAsset

logger = logging.getLogger(__name__)

# Scene labels that strongly indicate informational content
INFORMATIONAL_SCENES = {
    "document", "whiteboard", "text", "receipt", "menu", "sign",
    "poster", "screen", "monitor", "label", "book", "newspaper",
    "magazine", "letter", "note", "card", "ticket", "map",
    "billboard", "blackboard", "chalkboard",
}

# Scene labels that strongly indicate casual/personal content
CASUAL_SCENES = {
    "selfie", "portrait", "landscape", "sunset", "sunrise", "beach",
    "mountain", "sky", "pet", "cat", "dog", "food", "meal",
    "flower", "garden", "party", "celebration", "sport",
}


class Classification:
    INFORMATIONAL = "informational"
    CASUAL = "casual"
    AMBIGUOUS = "ambiguous"


def classify_image(asset: PhotoAsset, config: dict) -> str:
    """Classify a photo asset as informational or casual.

    Uses a tiered approach:
    1. Local heuristics (free, instant)
    2. Claude Code CLI fallback (for ambiguous cases)

    Returns: 'informational' or 'casual'
    """
    classification_config = config.get("classification", {})

    # Tier 1: Local heuristics
    result = _classify_local(asset, classification_config)

    if result != Classification.AMBIGUOUS:
        logger.info(f"Classified {asset.uuid} as {result} (local heuristics)")
        return result

    # Tier 2: Claude Code CLI fallback
    if classification_config.get("claude_fallback", True):
        result = _classify_with_claude(asset, config)
        logger.info(f"Classified {asset.uuid} as {result} (Claude CLI)")
        return result

    # Default to informational if no fallback (conservative — process rather than skip)
    logger.info(f"Classified {asset.uuid} as informational (default, no fallback)")
    return Classification.INFORMATIONAL


def _classify_local(asset: PhotoAsset, config: dict) -> str:
    """Classify using local heuristics from Apple Photos metadata."""
    # Rule 1: Screenshots are almost always informational
    if asset.is_screenshot and config.get("auto_process_screenshots", True):
        return Classification.INFORMATIONAL

    # Rule 2: Has text content detected by Apple → likely informational
    if asset.has_text:
        # But if it also has faces and no other informational signals, it might be a chat screenshot
        if asset.face_count > 0 and not asset.is_screenshot:
            return Classification.AMBIGUOUS
        return Classification.INFORMATIONAL

    # Rule 3: Scene labels
    informational_matches = set(asset.scene_labels) & INFORMATIONAL_SCENES
    casual_matches = set(asset.scene_labels) & CASUAL_SCENES

    if informational_matches and not casual_matches:
        return Classification.INFORMATIONAL

    if casual_matches and not informational_matches:
        return Classification.CASUAL

    # Rule 4: Faces only, no text → casual
    if asset.face_count > 0 and not asset.has_text and config.get("skip_faces_only", True):
        return Classification.CASUAL

    # Rule 5: No signals at all
    if not asset.scene_labels and not asset.has_text and asset.face_count == 0:
        return Classification.AMBIGUOUS

    return Classification.AMBIGUOUS


def _classify_with_claude(asset: PhotoAsset, config: dict | None = None) -> str:
    """Use Claude Code CLI to classify an ambiguous image."""
    if not asset.file_path:
        return Classification.INFORMATIONAL

    prompt = (
        "Look at this image. Is it INFORMATIONAL (contains text, schedule, URL, QR code, "
        "receipt, document, whiteboard, contact info, event details, or any useful data to remember) "
        "or CASUAL (selfie, family photo, scenery, food, pet, social moment)? "
        "Reply with exactly one word: INFORMATIONAL or CASUAL"
    )

    try:
        result = subprocess.run(
            ["claude", "-p", prompt, "--image", asset.file_path, "--output-format", "text"],
            capture_output=True,
            text=True,
            timeout=(config or {}).get("processing", {}).get("claude_timeout_classify", 30),
        )

        if result.returncode != 0:
            logger.warning(f"Claude CLI failed for {asset.uuid}: {result.stderr}")
            return Classification.INFORMATIONAL  # Default to processing

        response = result.stdout.strip().upper()
        if re.search(r"\bINFORMATIONAL\b", response):
            return Classification.INFORMATIONAL
        elif re.search(r"\bCASUAL\b", response):
            return Classification.CASUAL
        else:
            logger.warning(f"Unexpected Claude response for {asset.uuid}: {response}")
            return Classification.INFORMATIONAL

    except subprocess.TimeoutExpired:
        logger.warning(f"Claude CLI timed out for {asset.uuid}")
        return Classification.INFORMATIONAL
    except FileNotFoundError:
        logger.error("Claude CLI not found. Install Claude Code CLI.")
        return Classification.INFORMATIONAL

File: src/classifier.ts
import { execFile } from 'child_process';
import { promisify } from 'util';
import type { PhotoAsset } from './watcher.js';
import type { PicNoteConfig } from './config.js';

const execFileAsync = promisify(execFile);

export const Classification = {
  INFORMATIONAL: 'informational',
  CASUAL: 'casual',
  AMBIGUOUS: 'ambiguous',
} as const;

const INFORMATIONAL_SCENES = new Set([
  'document', 'whiteboard', 'text', 'receipt', 'menu', 'sign',
  'poster', 'screen', 'monitor', 'label', 'book', 'newspaper',
  'magazine', 'letter', 'note', 'card', 'ticket', 'map',
  'billboard', 'blackboard', 'chalkboard',
]);

const CASUAL_SCENES = new Set([
  'selfie', 'portrait', 'landscape', 'sunset', 'sunrise', 'beach',
  'mountain', 'sky', 'pet', 'cat', 'dog', 'food', 'meal',
  'flower', 'garden', 'party', 'celebration', 'sport',
]);

export async function classifyImage(asset: PhotoAsset, config: PicNoteConfig): Promise<string> {
  const classConfig = config.classification;

  const result = classifyLocal(asset, classConfig);
  if (result !== Classification.AMBIGUOUS) {
    return result;
  }

  if (classConfig.claude_fallback) {
    return classifyWithClaude(asset, config);
  }

  return Classification.INFORMATIONAL;
}

function classifyLocal(
  asset: PhotoAsset,
  config: PicNoteConfig['classification'],
): string {
  // Rule 1: Screenshots are almost always informational
  if (asset.isScreenshot && config.auto_process_screenshots) {
    return Classification.INFORMATIONAL;
  }

  // Rule 2: Has text → likely informational
  if (asset.hasText) {
    if (asset.faceCount > 0 && !asset.isScreenshot) {
      return Classification.AMBIGUOUS;
    }
    return Classification.INFORMATIONAL;
  }

  // Rule 3: Scene labels
  const infoMatches = asset.sceneLabels.filter((l) => INFORMATIONAL_SCENES.has(l));
  const casualMatches = asset.sceneLabels.filter((l) => CASUAL_SCENES.has(l));

  if (infoMatches.length > 0 && casualMatches.length === 0) {
    return Classification.INFORMATIONAL;
  }
  if (casualMatches.length > 0 && infoMatches.length === 0) {
    return Classification.CASUAL;
  }

  // Rule 4: Faces only, no text → casual
  if (asset.faceCount > 0 && !asset.hasText && config.skip_faces_only) {
    return Classification.CASUAL;
  }

  // Rule 5: No signals
  if (asset.sceneLabels.length === 0 && !asset.hasText && asset.faceCount === 0) {
    return Classification.AMBIGUOUS;
  }

  return Classification.AMBIGUOUS;
}

async function classifyWithClaude(asset: PhotoAsset, config: PicNoteConfig): Promise<string> {
  if (!asset.filePath) return Classification.INFORMATIONAL;

  const prompt =
    'Look at this image. Is it INFORMATIONAL (contains text, schedule, URL, QR code, ' +
    'receipt, document, whiteboard, contact info, event details, or any useful data to remember) ' +
    'or CASUAL (selfie, family photo, scenery, food, pet, social moment)? ' +
    'Reply with exactly one word: INFORMATIONAL or CASUAL';

  const timeout = config.processing.claude_timeout_classify * 1000;

  try {
    const { stdout } = await execFileAsync(
      'claude',
      ['-p', prompt, '--image', asset.filePath, '--output-format', 'text'],
      { timeout, encoding: 'utf-8' },
    );

    const response = stdout.trim().toUpperCase();
    if (/\bINFORMATIONAL\b/.test(response)) {
      return Classification.INFORMATIONAL;
    } else if (/\bCASUAL\b/.test(response)) {
      return Classification.CASUAL;
    }
    console.warn(`Unexpected Claude response for ${asset.uuid}: ${response}`);
    return Classification.INFORMATIONAL;
  } catch (err: any) {
    console.warn(`Claude CLI failed for ${asset.uuid}: ${err.message}`);
    return Classification.INFORMATIONAL;
  }
}
